src/libopensc/pkcs15-piv.c

```diff
@@ -968,17 +986,17 @@ static int sc_pkcs15emu_piv_init(sc_pkcs15_card_t *p15card)
 		sc_format_path(pins[i].path, &pin_info.path);
 
 		label = pins[i].label;
-		if (i == 0 &&
+		if ((i == 0 || pins[i].cardmod) &&
 			sc_card_ctl(card, SC_CARDCTL_PIV_PIN_PREFERENCE,
 					&pin_ref) == 0 &&
 				pin_ref == 0x00) { /* must be 80 for PIV pin, or 00 for Global PIN */
 			pin_info.attrs.pin.reference = pin_ref;
-			pin_info.attrs.pin.flags &= ~SC_PKCS15_PIN_FLAG_LOCAL;
 			label = "Global PIN";
 		}
+
 		strncpy(pin_obj.label, label, SC_PKCS15_MAX_LABEL_SIZE - 1);
 		pin_obj.flags = pins[i].obj_flags;
-		if (i == 0 && pin_info.attrs.pin.reference == 0x80) {
+		if ((i == 0 || pins[i].cardmod)) {
```

Comment:
This seems to change the logic condition from (`i == 0` **AND** `pin.reference == 0x80`) to (`i == 0` **OR** `cardmod`).
Is this change intentional?

Reply:
It logic may need some changes.  The intent is when run by minidriver, the pins[i].cardmod is set on pin[2] which will be treated as the Sign PIN by minisriver with the same pin reference as PIN[0].  With PKCS11 or any application other than minidriver, lines 969-971 will cut the loop short and only two pins are defined: pin[0] as the auth pin, and pin[1] as the PUK PIN/Admin PIN. 

`((i == 0 || pins[i].cardmod)) {`  is really `if (i == 0 || i == 2)` set the auth_id of the PUK/ADMIN PIN on the other pin(s).
sp800-73-X (all version) only defines one PIN which when used with the 9C key, must be verified just before the APDU (00 87 XX 9C) crypto operation.       

sp800-73-X also allows for a Global PIN  with reference 00 or a local pin with reference 80 to be use for pin[0] It defines which is preferred or which is the only one to use from the DISCOVERY object.  (The prompt to the user
will say USER PIN or Global PIN, just incase they are different.   

NIST says only the local PUK PIN (key_ref 81) can be used by the user to unlock/change the pin. Some PIV like cards may allow a Global PUK/Admin pin (key_ref 01) to be used as a PUK. This restriction was relaxed by removing `pin_info.attrs.pin.reference == 0x80`   The card should enforce this.    

The point is when not run with minidriver, pkcs15 tools and routines will list only two pins. Only the minidriver sees a copy of the pin[0] as pin[2] so minidriver uses pin[2] when the 9C key is being used: `usrr_consent > 0`. The minidriver tests for `need_pin_always`.